src/api/worker/search/Indexer.js

```diff
@@ -419,61 +426,78 @@ export class Indexer {
+					for (let batch of eventBatchesOnServer) {
+						const batchId = getElementId(batch)
+						if (groupIdToEventBatch.eventBatchIds.indexOf(batchId) === -1
+							&& firstBiggerThanSecond(batchId, startId)) {
+							batchesToQueue.push({groupId: groupIdToEventBatch.groupId, batchId, events: batch.events})
+							const lastBatch = lastLoadedBatchIdInGroup.get(groupIdToEventBatch.groupId)
+							if (lastBatch == null || firstBiggerThanSecond(batchId, lastBatch)) {
+								lastLoadedBatchIdInGroup.set(groupIdToEventBatch.groupId, batchId)
+							}
+						}
+					}
+					// Good scenario: we know when we stopped, we can process events we did not process yet and catch up the server
+					//
+					//
+					// [4, 3, 2, 1]                          - processed events, lastBatchId =1
+					// load from lowest id 1 -1
+					// [0.9, 1, 2, 3, 4, 5, 6, 7, 8]         - last X events from server
+					// => [5, 6, 7, 8]                       - batches to queue
+					//
+					// Bad scenario: we don' know where we stopped, server doesn't have events to fill the gap anymore, we cannot fix the index.
+					// [4, 3, 2, 1] - processed events, lastBatchId = 1
+					// [7, 5, 9, 10] - last events from server
+					// => [7, 5, 9, 10] - batches to queue - nothing has been processed before so we are out of sync
+
+					// We only want to do this check for clients that haven't yet saved the index time
+					// This can be removed in the future
+					if (lastIndexTimeMs == null && eventBatchesOnServer.length === batchesToQueue.length) {
+						// Bad scenario happened.
+						// None of the events we want to process were processed before, we're too far away, stop the process and delete
+						// the index.
+						throw new OutOfSyncError(`We lost entity events for group ${groupIdToEventBatch.groupId}. start id was ${startId}`)
+					}
+					batchesOfAllGroups.push(...batchesToQueue)
 				}
-			})
-			.then(() => {
-				// add all batches of all groups in one step to avoid that just some groups are added when a ServiceUnavailableError occurs
-				// Add them directly to the core so that they are added before the realtime batches
-				this._core.addBatchesToQueue(batchesOfAllGroups)
-				this._initiallyLoadedBatchIdsPerGroup = lastLoadedBatchIdInGroup
-				this._realtimeEventQueue.resume()
-				this.startProcessing()
-			})
+			}
+		} catch (e) {
+			if (e instanceof NotAuthorizedError) {
+				console.log("could not download entity updates => lost permission on list")
+				return
+			}
+			throw e
+		}
+
+		// add all batches of all groups in one step to avoid that just some groups are added when a ServiceUnavailableError occurs
+		// Add them directly to the core so that they are added before the realtime batches
+		this._core.addBatchesToQueue(batchesOfAllGroups)
+		this._initiallyLoadedBatchIdsPerGroup = lastLoadedBatchIdInGroup
+		this._realtimeEventQueue.resume()
+		this.startProcessing()
+		await this._writeServerTimestamp()
```

Comment:
We probably cannot write it earlier, right?
It's just often the case with big mailboxes that we can't index fast enough and you log out but maybe that's another issue.